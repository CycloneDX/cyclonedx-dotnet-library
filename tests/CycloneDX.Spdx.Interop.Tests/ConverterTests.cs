// This file is part of CycloneDX Library for .NET
//
// Licensed under the Apache License, Version 2.0 (the “License”);
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
// http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an “AS IS” BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
//
// SPDX-License-Identifier: Apache-2.0
// Copyright (c) OWASP Foundation. All Rights Reserved.

using System;
using System.IO;
using Xunit;
using Snapshooter;
using Snapshooter.Xunit;
using CycloneDX.Models;
using CycloneDX.Spdx.Serialization;
using System.Reflection.Metadata;

namespace CycloneDX.Spdx.Interop.Tests
{
    public class ConverterTests
    {
        [Theory]
        [InlineData("v2.2")]
        [InlineData("v2.3")]
        public void FromSpdxToCDXToSpdxRoundTripTest(string version)
        {
            var resourceFilename = Path.Join("Resources", "Spdx", version, "document" + ".json");
            var fileContents = File.ReadAllText(resourceFilename);

            var spdxDocument = JsonSerializer.Deserialize(fileContents);
            var cdxBom = spdxDocument.ToCycloneDX();
            var result = cdxBom.ToSpdx();
            var resultString = JsonSerializer.Serialize(result);

            Snapshot.Match(resultString, SnapshotNameExtension.Create($"{version}document"));
        }

        [Theory]
        [InlineData("v2.2")]
        [InlineData("v2.3")]
        public void FromSpdxToCDXTest(string version)
        {
            var resourceFilename = Path.Join("Resources", "Spdx",version, "document" + ".json");
            var fileContents = File.ReadAllText(resourceFilename);

            var spdxDocument = JsonSerializer.Deserialize(fileContents);
            var cdxBom = spdxDocument.ToCycloneDX();
            var resultString = CycloneDX.Json.Serializer.Serialize(cdxBom);

            Snapshot.Match(resultString, SnapshotNameExtension.Create($"{version}document"));
        }

        [Theory]
        [InlineData("assembly")]
        [InlineData("bom")]
        [InlineData("component-hashes")]
        [InlineData("component-ref")]
        [InlineData("component-swid")]
        [InlineData("component-swid-full")]
        [InlineData("component-types")]
        [InlineData("compositions")]
        [InlineData("dependency")]
        [InlineData("empty-components")]
        [InlineData("evidence")]
        [InlineData("external-reference")]
        [InlineData("license-expression")]
        [InlineData("license-id")]
        [InlineData("license-name")]
        [InlineData("metadata-author")]
        [InlineData("metadata-license")]
        [InlineData("metadata-manufacture")]
        [InlineData("metadata-supplier")]
        [InlineData("metadata-timestamp")]
        [InlineData("metadata-tool")]
        [InlineData("minimal-viable")]
        [InlineData("patch")]
        [InlineData("properties")]
        [InlineData("service")]
        [InlineData("service-empty-objects")]
        public void FromCDXToSpdxTest(string baseFilename)
        {
            var resourceFilename = Path.Join("Resources", "CycloneDX", "v1.3", $"valid-{baseFilename}-1.3.json");
            var fileContents = File.ReadAllText(resourceFilename);

            var cdxBom = CycloneDX.Json.Serializer.Deserialize(fileContents);
            // set the BOM timestamp if it hasn't been already
            // it's required for converting to SPDX
            // and if we don't set it the tests will fail because the time has changed from last test run
            cdxBom.Metadata = cdxBom.Metadata ?? new Metadata();
            cdxBom.Metadata.Timestamp = cdxBom.Metadata.Timestamp ?? new DateTime(2021, 12, 24, 12, 00, 00, DateTimeKind.Utc);

            var spdxDocument = cdxBom.ToSpdx();
            var resultString = JsonSerializer.Serialize(spdxDocument);

            Snapshot.Match(resultString, SnapshotNameExtension.Create(baseFilename));
        }

        [Theory]
        [InlineData("assembly")]
        [InlineData("bom")]
        [InlineData("component-hashes")]
        [InlineData("component-ref")]
        [InlineData("component-swid")]
        [InlineData("component-swid-full")]
        [InlineData("component-types")]
        [InlineData("compositions")]
        [InlineData("dependency")]
        [InlineData("empty-components")]
        [InlineData("evidence")]
        [InlineData("external-reference")]
        [InlineData("license-expression")]
        [InlineData("license-id")]
        [InlineData("license-name")]
        [InlineData("metadata-author")]
        [InlineData("metadata-license")]
        [InlineData("metadata-manufacture")]
        [InlineData("metadata-supplier")]
        [InlineData("metadata-timestamp")]
        [InlineData("metadata-tool")]
        [InlineData("minimal-viable")]
        [InlineData("patch")]
        [InlineData("properties")]
        [InlineData("service")]
        [InlineData("service-empty-objects")]
        public void FromCDXToSpdxToCDXRoundTripTest(string baseFilename)
        {
            var resourceFilename = Path.Join("Resources", "CycloneDX", "v1.3", $"valid-{baseFilename}-1.3.json");
            var fileContents = File.ReadAllText(resourceFilename);

            var cdxBom = CycloneDX.Json.Serializer.Deserialize(fileContents);
            // set the BOM timestamp if it hasn't been already
            // it's required for converting to SPDX
            // and if we don't set it the tests will fail because the time has changed from last test run
            cdxBom.Metadata = cdxBom.Metadata ?? new Metadata();
            cdxBom.Metadata.Timestamp = cdxBom.Metadata.Timestamp ?? new DateTime(2021, 12, 24, 12, 00, 00, DateTimeKind.Utc);

            var spdxDocument = cdxBom.ToSpdx();
            var resultBom = spdxDocument.ToCycloneDX();
            var resultString = CycloneDX.Json.Serializer.Serialize(resultBom);

            Snapshot.Match(resultString, SnapshotNameExtension.Create(baseFilename));
        }
    }
}
