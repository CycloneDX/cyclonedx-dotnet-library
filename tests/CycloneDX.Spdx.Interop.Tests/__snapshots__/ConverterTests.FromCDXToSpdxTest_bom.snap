{
  "SPDXID": "SPDXRef-DOCUMENT",
  "spdxVersion": "SPDX-2.3",
  "creationInfo": {
    "comment": "This SPDX document has been converted from CycloneDX format.",
    "created": "2020-04-13T20:20:39Z",
    "creators": [
      "Tool: Awesome Tool-9.1.2",
      "Person: Samantha Wright (samantha.wright@example.com)"
    ]
  },
  "name": "Acme Application-9.1.1",
  "dataLicense": "CC0-1.0",
  "documentNamespace": "http://spdx.org/spdxdocs/Acme Application-9.1.1-3e671687-395b-41f5-a30f-a58921a69b79",
  "packages": [
    {
      "SPDXID": "SPDXRef-pkg:npm/acme/component@1.0.0",
      "checksums": [
        {
          "algorithm": "MD5",
          "checksumValue": "3942447fac867ae5cdb3229b658f4d48"
        },
        {
          "algorithm": "SHA1",
          "checksumValue": "e6b1000b94e835ffd37f4c6dcbdad43f4b48a02a"
        },
        {
          "algorithm": "SHA256",
          "checksumValue": "f498a8ff2dd007e29c2074f5e4b01a9a01775c3ff3aeaf6906ea503bc5791b7b"
        },
        {
          "algorithm": "SHA512",
          "checksumValue": "e8f33e424f3f4ed6db76a482fde1a5298970e442c531729119e37991884bdffab4f9426b7ee11fccd074eeda0634d71697d6f88a460dce0ac8d627a29f7d1282"
        }
      ],
      "downloadLocation": "NOASSERTION",
      "licenseDeclared": "Apache-2.0",
      "name": "tomcat-catalina",
      "primaryPackagePurpose": "LIBRARY",
      "versionInfo": "9.0.14"
    },
    {
      "SPDXID": "SPDXRef-Package-2",
      "downloadLocation": "NOASSERTION",
      "name": "mylibrary",
      "originator": "Person: Example Super Heros ()",
      "primaryPackagePurpose": "LIBRARY",
      "supplier": "Person: Example, Inc. (support@example.com)",
      "versionInfo": "1.0.0"
    }
  ]
}
